File: apps/server/src/modules/cicd/cicd.service.ts
import { prisma } from "../../lib/prisma";
import { logger } from "../../lib/logger";
import { notFound } from "../../lib/errors";
import { emitter } from "../../lib/emitter";
import type { CicdStageLog } from "../../lib/socket.events";

// ── Helpers ───────────────────────────────────────────────────────────────────

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

function deterministicPass(seed: string, failRate: number): boolean {
  let hash = 0;
  for (let i = 0; i < seed.length; i++) hash = (hash * 31 + seed.charCodeAt(i)) >>> 0;
  return (hash % 100) >= failRate * 100;
}

function generatePreviewUrl(deploymentId: string): string {
  return `https://preview-${deploymentId.slice(-8)}.adw-deploy.example.com`;
}

function emitUpdate(
  deploymentId: string,
  projectId: string,
  taskId: string | null,
  status: "PENDING" | "RUNNING" | "SUCCESS" | "FAILED",
  log: CicdStageLog[],
  extra: { stage?: string; previewUrl?: string; errorMsg?: string } = {},
) {
  emitter.deploymentUpdated({
    deploymentId, projectId, taskId, status, log,
    updatedAt: new Date().toISOString(),
    ...extra,
  });
}

interface StageConfig {
  name: string;
  failRate: number;
  minMs: number;
  maxMs: number;
  passDetail: string;
  failDetail: string;
  dbField?: "testDurationMs" | "buildDurationMs";
}

const STAGES: StageConfig[] = [
  {
    name: "tests",
    failRate: 0.1,
    minMs: 1500,
    maxMs: 2500,
    passDetail: "All tests passed",
    failDetail: "2 tests failed — assertion error in handler.test.ts",
    dbField: "testDurationMs",
  },
  {
    name: "build",
    failRate: 0.05,
    minMs: 2000,
    maxMs: 3500,
    passDetail: "Build succeeded — 0 errors, 0 warnings",
    failDetail: "TypeScript error: Type 'string' is not assignable to type 'number'",
    dbField: "buildDurationMs",
  },
  {
    name: "deploy",
    failRate: 0,
    minMs: 800,
    maxMs: 1400,
    passDetail: "", // set dynamically
    failDetail: "",
  },
];

// ── Pipeline ──────────────────────────────────────────────────────────────────

export async function runCicdPipeline(projectId: string, taskId: string | null): Promise<void> {
  const project = await prisma.project.findUnique({ where: { id: projectId }, select: { id: true } });
  if (!project) throw notFound("Project");

  const seed = `${projectId}${taskId ?? ""}`;
  const deployment = await prisma.deployment.create({
    data: { projectId, taskId, status: "RUNNING", log: "[]" },
  });
  const { id: deploymentId } = deployment;
  const log: CicdStageLog[] = [];

  logger.info("CI/CD pipeline started", { deploymentId, projectId, taskId });

  for (let i = 0; i < STAGES.length; i++) {
    const stage = STAGES[i]!;
    const durationMs = stage.minMs + Math.floor(Math.random() * (stage.maxMs - stage.minMs));

    // Mark stage as running
    log[i] = { stage: stage.name, status: "running" };
    await prisma.deployment.update({ where: { id: deploymentId }, data: { log: JSON.stringify(log) } });
    emitUpdate(deploymentId, projectId, taskId, "RUNNING", log, { stage: stage.name });

    await sleep(durationMs);

    const passed = deterministicPass(seed + stage.name, stage.failRate);

    if (stage.name === "deploy") {
      const previewUrl = generatePreviewUrl(deploymentId);
      log[i] = { stage: stage.name, status: "passed", durationMs, detail: `Preview deployed to ${previewUrl}` };
      await prisma.deployment.update({
        where: { id: deploymentId },
        data: { status: "SUCCESS", previewUrl, log: JSON.stringify(log) },
      });
      emitUpdate(deploymentId, projectId, taskId, "SUCCESS", log, { previewUrl });
      logger.info("CI/CD pipeline succeeded", { deploymentId, previewUrl });
      return;
    }

    log[i] = {
      stage: stage.name,
      status: passed ? "passed" : "failed",
      durationMs,
      detail: passed ? stage.passDetail : stage.failDetail,
    };

    const dbUpdate: Record<string, unknown> = { log: JSON.stringify(log) };
    if (stage.dbField) dbUpdate[stage.dbField] = durationMs;
    await prisma.deployment.update({ where: { id: deploymentId }, data: dbUpdate });

    if (!passed) {
      await prisma.deployment.update({
        where: { id: deploymentId },
        data: { status: "FAILED", errorMsg: log[i]!.detail },
      });
      emitUpdate(deploymentId, projectId, taskId, "FAILED", log, { errorMsg: log[i]!.detail });
      logger.warn(`CI/CD failed at ${stage.name}`, { deploymentId });
      return;
    }

    emitUpdate(deploymentId, projectId, taskId, "RUNNING", log, { stage: stage.name });
  }
}

// ── Queries ───────────────────────────────────────────────────────────────────

export async function listDeployments(projectId: string) {
  const project = await prisma.project.findUnique({ where: { id: projectId }, select: { id: true } });
  if (!project) throw notFound("Project");

  const rows = await prisma.deployment.findMany({
    where: { projectId },
    orderBy: { createdAt: "desc" },
    take: 50,
  });
  return rows.map((d) => ({ ...d, log: JSON.parse(d.log) as CicdStageLog[] }));
}

export async function getDeployment(id: string) {
  const d = await prisma.deployment.findUnique({ where: { id } });
  if (!d) throw notFound("Deployment");
  return { ...d, log: JSON.parse(d.log) as CicdStageLog[] };
}

File: apps/server-py/src/modules/cicd/cicd_service.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional

from src.core.database import db
from src.core.errors import not_found
from src.core.utils import now_iso
from src.realtime import emitter
from src.realtime.events import DeploymentUpdatedPayload, CicdStageLog
from src.core.logger import logger
from src.auth.auth import AuthUser, get_current_user, log_action

import asyncio
import random

router = APIRouter()

STAGES = [
    {"name": "tests",  "fail_rate": 0.1,  "min_ms": 1500, "max_ms": 2500,
     "pass_detail": "All tests passed",
     "fail_detail": "2 tests failed — assertion error in handler_test.py",
     "logs": [
         "🔍 Discovering test files...",
         "📦 Installing test dependencies...",
         "🧪 Running pytest...",
         "✅ test_auth.py passed",
         "✅ test_projects.py passed",
         "✅ test_tasks.py passed",
     ]},
    {"name": "build",  "fail_rate": 0.05, "min_ms": 2000, "max_ms": 3500,
     "pass_detail": "Build succeeded — 0 errors, 0 warnings",
     "fail_detail": "TypeError: unsupported operand type(s) for +: 'int' and 'str'",
     "logs": [
         "🔧 Compiling source files...",
         "📦 Bundling dependencies...",
         "🗜 Optimizing assets...",
         "✅ Build completed successfully",
     ]},
    {"name": "deploy", "fail_rate": 0.0,  "min_ms": 800,  "max_ms": 1400,
     "pass_detail": "", "fail_detail": "",
     "logs": [
         "🚀 Preparing deployment package...",
         "📡 Connecting to platform...",
         "⬆ Uploading files...",
     ]},
]


def _deterministic_pass(seed: str, fail_rate: float) -> bool:
    h = 0
    for c in seed:
        h = (h * 31 + ord(c)) & 0xFFFFFFFF
    return (h % 100) >= int(fail_rate * 100)


def _stage_logs(log: list[dict]) -> list[CicdStageLog]:
    return [CicdStageLog(**e) for e in log]


async def run_cicd_pipeline(project_id: str, task_id: Optional[str], platform: Optional[str] = None) -> None:
    if not db.collection("projects").document(project_id).get().exists:
        raise not_found("Project")

    seed = f"{project_id}{task_id or ''}"
    platform_name = platform or "default"

    deploy_ref = db.collection("deployments").add({
        "projectId": project_id, "taskId": task_id,
        "platform": platform_name,
        "status": "RUNNING", "log": [],
        "createdAt": now_iso(), "updatedAt": now_iso(),
    })[1]
    deployment_id = deploy_ref.id
    log: list[dict] = []

    logger.info(f"CI/CD pipeline started: deployment={deployment_id} platform={platform_name}")
    await emitter.emit_cicd_log(project_id, deployment_id, "pipeline", f"🚀 Pipeline started for {platform_name.title()} deployment")

    for stage in STAGES:
        duration_ms = random.randint(stage["min_ms"], stage["max_ms"])
        step_delay = (duration_ms / 1000) / len(stage["logs"])

        log.append({"stage": stage["name"], "status": "running"})
        deploy_ref.update({"log": log, "updatedAt": now_iso()})
        await emitter.emit_deployment_updated(DeploymentUpdatedPayload(
            deploymentId=deployment_id, projectId=project_id, taskId=task_id,
            status="RUNNING", stage=stage["name"],
            log=_stage_logs(log), updatedAt=now_iso(),
        ))

        # Stream each log line with delay
        for line in stage["logs"]:
            await asyncio.sleep(step_delay)
            await emitter.emit_cicd_log(project_id, deployment_id, stage["name"], line)

        if stage["name"] == "deploy":
            if platform == "vercel":
                preview_url = f"https://adw-{deployment_id[-8:]}.vercel.app"
            elif platform == "github":
                preview_url = f"https://username.github.io/adw-{deployment_id[-8:]}"
            elif platform == "infinityfree":
                preview_url = f"http://adw-{deployment_id[-8:]}.rf.gd"
            else:
                preview_url = f"https://preview-{deployment_id[-8:]}.adw-deploy.example.com"

            await emitter.emit_cicd_log(project_id, deployment_id, stage["name"], f"✅ Deployed → {preview_url}")
            log[-1] = {"stage": "deploy", "status": "passed", "durationMs": duration_ms,
                       "detail": f"Deployed to {platform_name.title()} → {preview_url}"}
            deploy_ref.update({"status": "SUCCESS", "previewUrl": preview_url, "log": log, "updatedAt": now_iso()})
            await emitter.emit_deployment_updated(DeploymentUpdatedPayload(
                deploymentId=deployment_id, projectId=project_id, taskId=task_id,
                status="SUCCESS", previewUrl=preview_url,
                log=_stage_logs(log), updatedAt=now_iso(),
            ))
            await emitter.emit_cicd_log(project_id, deployment_id, "pipeline", "🎉 Pipeline completed successfully!")
            logger.info(f"CI/CD succeeded: deployment={deployment_id} preview={preview_url}")
            return

        passed = _deterministic_pass(seed + stage["name"], stage["fail_rate"])
        detail = stage["pass_detail"] if passed else stage["fail_detail"]
        log[-1] = {"stage": stage["name"], "status": "passed" if passed else "failed",
                   "durationMs": duration_ms, "detail": detail}
        deploy_ref.update({"log": log, "updatedAt": now_iso()})

        if not passed:
            await emitter.emit_cicd_log(project_id, deployment_id, stage["name"], f"❌ {detail}", level="error")
            await emitter.emit_cicd_log(project_id, deployment_id, "pipeline", f"💥 Pipeline failed at {stage['name']} stage", level="error")
            deploy_ref.update({"status": "FAILED", "errorMsg": detail, "updatedAt": now_iso()})
            await emitter.emit_deployment_updated(DeploymentUpdatedPayload(
                deploymentId=deployment_id, projectId=project_id, taskId=task_id,
                status="FAILED", errorMsg=detail,
                log=_stage_logs(log), updatedAt=now_iso(),
            ))
            logger.warning(f"CI/CD failed at {stage['name']}: deployment={deployment_id}")
            return

        await emitter.emit_cicd_log(project_id, deployment_id, stage["name"], f"✅ {detail}")
        await emitter.emit_deployment_updated(DeploymentUpdatedPayload(
            deploymentId=deployment_id, projectId=project_id, taskId=task_id,
            status="RUNNING", stage=stage["name"],
            log=_stage_logs(log), updatedAt=now_iso(),
        ))


class TriggerCicdRequest(BaseModel):
    projectId: str
    taskId: Optional[str] = None
    platform: Optional[str] = None


@router.post("/deploy")
async def trigger_cicd(body: TriggerCicdRequest, user: AuthUser = Depends(get_current_user)):
    project_doc = db.collection("projects").document(body.projectId).get()
    if not project_doc.exists:
        raise not_found("Project")

    project_data = project_doc.to_dict()
    if not user.can_access_resource(project_data.get("ownerId")):
        raise HTTPException(status_code=403, detail="Access denied")

    asyncio.create_task(run_cicd_pipeline(body.projectId, body.taskId, body.platform))
    await log_action(user, "CICD_TRIGGER", {"projectId": body.projectId, "taskId": body.taskId, "platform": body.platform or "default"})
    return {"message": f"CI/CD pipeline triggered for {body.platform or 'default'} deployment"}


@router.get("/deployments")
async def list_deployments(projectId: str, user: AuthUser = Depends(get_current_user)):
    project_doc = db.collection("projects").document(projectId).get()
    if not project_doc.exists:
        raise not_found("Project")

    if not user.can_access_resource(project_doc.to_dict().get("ownerId")):
        raise HTTPException(status_code=403, detail="Access denied")

    snap = (
        db.collection("deployments")
        .where("projectId", "==", projectId)
        .order_by("createdAt", direction="DESCENDING")
        .limit(50)
        .stream()
    )
    return [{"id": d.id, **d.to_dict()} for d in snap]


@router.get("/deployments/{deployment_id}")
async def get_deployment(deployment_id: str, user: AuthUser = Depends(get_current_user)):
    doc = db.collection("deployments").document(deployment_id).get()
    if not doc.exists:
        raise not_found("Deployment")

    deployment_data = doc.to_dict()
    project_id = deployment_data.get("projectId")
    if project_id:
        project_doc = db.collection("projects").document(project_id).get()
        if project_doc.exists and not user.can_access_resource(project_doc.to_dict().get("ownerId")):
            raise HTTPException(status_code=403, detail="Access denied")

    return {"id": doc.id, **deployment_data}
